### ISSUE TITLE
QUESTION: woocommerce product search results sort by stock_status

### ISSUE THREAD

**Issue by beazuadmin** [10up/ElasticPress#2473]
Is there a way to use _stock_status metadata in a woocommerce product to sort search results such that it shows all out-of-stock products last? 

When ElasticPress is not active, the following filter works correctly:
```php
add_filter( 'woocommerce_get_catalog_ordering_args', 'show_instock_first', 9999 );
function show_instock_first( $args ) {
   $args['orderby'] = 'meta_value';
   $args['order'] = 'ASC';
   $args['meta_key'] = '_stock_status';
   return $args;
}
```

But ElaticPress seems to ignore that. 

If I step on the function get_es_posts in includes/classes/Indexable/Post/QueryIntegration.php I can see that the $query is setup correctly, and it has the right _stock_status meta_query set. However, it seems that $query->query_vars has 'orderby' set to false and elastic press never picks up the _stock_status meta query from the query. 

Any idea how to make this work? 
Thanks

**Comment by beazuadmin** [comment]
Just another note. I found that the Woocommerce Feature in ElasticPress is overriding the orderby value in the `public function translate_args( $query )` (includes/classes/Feature/WooCommerce/WooCommerce.php)

If I comment the following lines: 
```php
			if ( ! empty( $s ) ) {
				**// $query->set( 'orderby', false ); // Just order by relevance.** 

				/**
				 * Default order when doing search in Woocommerce is 'ASC'
				 * These lines will change it to 'DESC' as we want to most relevant result
				 */
				if ( empty( $_GET['orderby'] ) && $query->is_main_query() ) { // phpcs:ignore WordPress.Security.NonceVerification
					**// $query->set( 'order', 'DESC' );**
				}
```
Then the search results come ordered as intended (ie, in stock products come before out of stock). 

But I can see that by doing the ASC order I might be messing with the relevancy of the result. 

Can anyone suggest what is the best way to achieve this? 
Thanks!

**Comment by felipeelia** [comment]
Hi @beazuadmin,

First of all, sorry for the late reply here. The lines you commented out don't seem to be needed anymore, as I outlined in #2577.  With those commented out, you can use this code to have the order by stock AND relevance applied:

```
function show_instock_first( $args ) {
	$args['orderby'] = [
		'meta._stock_status.raw' => 'ASC',
		'relevance'              => 'DESC',
	];

	return $args;
}
add_filter( 'woocommerce_get_catalog_ordering_args', 'show_instock_first', 9999 );
```

That PR will go through our code review process and will probably get merged for our next release. Until then, for you and any other users facing the same issue, this snippet should be enough:

```
add_action(
	'pre_get_posts',
	function( $query ) {
		if ( 'product_query' !== $query->get( 'wc_query' ) ) {
			return;
		}

		$query->set(
			'orderby',
			[
				'meta._stock_status.raw' => 'ASC',
				'relevance'              => 'DESC',
			]
		);
	},
	99
);
```

Thanks!

**Comment by felipeelia** [comment]
As #2577 was merged, I'm closing out the issue.

**Comment by beazuadmin** [comment]
Hi @felipeelia  just to confirm that the temporary solution you outline above works with the current version of the plugin in production. Thanks!

**Comment by Ahammed-Navas** [comment]
Hello @felipeelia 

The code worked. Now the out-of-stock products are showing at the end of my listing. but when I choose the default order status as “price – low to high”, it does not work. Would you know any workaround to make it work?

Thank you.

**Comment by Ahammed-Navas** [comment]
Hello @felipeelia 

Is there any update on the above issue?

Thanks

**Comment by Ahammed-Navas** [comment]
Hello @felipeelia 

Thanks for your reply.

I tried adding the code above you shared and unfortunately, it did not work as expected. Sorting now works correctly. However, sorting out-of-stock products last in the archive page is not working. Can you please provide us with the correct code? We need to resolve the issue as soon as possible.

Thank you

**Comment by brandwaffle** [comment]
Hi @Ahammed-Navas my colleague @felipeelia has very nicely explained to you that this work is custom development well outside the scope of our free, open source plugin support that we provide here for bug fixes and feature requests.

If you would like to someone else to write code for you, we can provide custom consulting on a paid basis. You may also prefer to use Codeable which has several ElasticPress experts available: https://www.codeable.io/developers/?developer_search=elasticpress

**Comment by Ahammed-Navas** [comment]
Hi @brandwaffle ,
Okay, But When ElasticPress plugin is not active on the site, the following filter works correctly:

`function order_by_stock_status($posts_clauses) {
    global $wpdb;
    // only change query on WooCommerce loops
    if (is_woocommerce() && (is_shop() || is_product_category() || is_product_tag() || is_product_taxonomy())) {
        $posts_clauses['join'] .= " INNER JOIN $wpdb->postmeta istockstatus ON ($wpdb->posts.ID = istockstatus.post_id) ";
        $posts_clauses['orderby'] = " istockstatus.meta_value ASC, " . $posts_clauses['orderby'];
        $posts_clauses['where'] = " AND istockstatus.meta_key = '_stock_status' AND istockstatus.meta_value <> '' " . $posts_clauses['where'];
    }
    return $posts_clauses;
}`

Therefore, the issue is due to ElasticPress. That's why I asked you about it. Would you mind checking the code again? I hope we can resolve it.

Thanks

**Comment by brandwaffle** [comment]
@Ahammed-Navas it works without ElasticPress because you have written a direct SQL query and as we've explained to you before, ElasticPress integrates with WP_Query. ElasticPress doesn't support 100% of all themes and plugins in the WordPress ecosystem, and it certainly can't support custom code that circumvents the WP Query API, because that's impossible. I'm sorry you're not able to make it work for your purposes, but we do not provide custom development support here, as I've stated numerous times.

If you have legitimate issues to share on this repo, you are welcome to do so. However, if you continue to request individualized support for your specific integration, we may have to take additional action up to and including blocking your ability to submit tickets to our repo.
